test/Microsoft.AspNet.Mvc.Core.Test/SessionStateTempDataProviderTest.cs

```diff
@@ -76,6 +77,144 @@ public void Save_NullSession_NonEmptyDictionary_Throws()
+        }
+
+        [Theory]
+        [MemberData(nameof(InvalidTypes))]
+        public void EnsureObjectCanBeSerialized_InvalidType_Throws(object value, Type type)
+        {
+            // Arrange
+            var testProvider = new SessionStateTempDataProvider();
+
+            // Act & Assert
+            var exception = Assert.Throws<InvalidOperationException>(() =>
+            {
+                testProvider.EnsureObjectCanBeSerialized(value);
+            });
+            Assert.Equal($"The type {type} cannot be serialized to Session by '{typeof(SessionStateTempDataProvider).FullName}'.",
+                exception.Message);
+        }
+
+        public static TheoryData<object, Type> InvalidDictionaryTypes
+        {
+            get
+            {
+                return new TheoryData<object, Type>
+                {
+                    { new Dictionary<int, string>(), typeof(int) },
+                    { new Dictionary<Uri, Guid>(), typeof(Uri) },
+                    { new Dictionary<object, string>(), typeof(object) },
+                    { new Dictionary<TestItem, TestItem>(), typeof(TestItem) }
+                };
+            }
+        }
+
+        [Theory]
+        [MemberData(nameof(InvalidDictionaryTypes))]
+        public void EnsureObjectCanBeSerialized_InvalidDictionaryType_Throws(object value, Type type)
+        {
+            // Arrange
+            var testProvider = new SessionStateTempDataProvider();
+
+            // Act & Assert
+            var exception = Assert.Throws<InvalidOperationException>(() =>
+            {
+                testProvider.EnsureObjectCanBeSerialized(value);
+            });
+            Assert.Equal($"The dictionary with TKey {type} cannot be serialized to Session by '{typeof(SessionStateTempDataProvider).FullName}'.",
+                exception.Message);
+        }
+
+        public static TheoryData<object> ValidTypes
+        {
+            get
+            {
+                return new TheoryData<object>
+                {
+                    { 10 },
+                    { new int[]{ 10, 20 } },
+                    { "FooValue" },
+                    { new Uri("http://Foo") },
+                    { Guid.NewGuid() },
+                    { new List<string> { "foo", "bar" } },
+                    { new DateTimeOffset() },
+                    { 100.1m },
+                    { new Dictionary<string, int>() },
+                    { new Uri[] { new Uri("http://Foo"), new Uri("http://Bar") } }
+                };
+            }
+        }
+
+        [Theory]
+        [MemberData(nameof(ValidTypes))]
+        public void EnsureObjectCanBeSerialized_ValidType_DoesNotThrow(object value)
+        {
+            // Arrange
+            var testProvider = new SessionStateTempDataProvider();
+
+            // Act & Assert (Does not throw)
+            testProvider.EnsureObjectCanBeSerialized(value);
+        }
+
+        [Fact]
+        public void SaveAndLoad_WorksAsExpected()
```

Comment:
Maybe call this `SimpleTypesCanBeStoredAndLoaded` - it's slightly less vague than AsExpected.